# Labolatorium 5 -  Pandas — Series, DataFrame, eksploracja danych

## Ćwiczenie 1
Series — tworzenie i operacje

In [1]:
import pandas as pd
import numpy as np

In [2]:
oceny = pd.Series([4.5, 3.0, 5.0, 3.5, 4.0])
print(oceny)

0    4.5
1    3.0
2    5.0
3    3.5
4    4.0
dtype: float64


In [24]:
pensje = {'Anna': 5500, 'Jan': 7200, 'Ewa': 4800, 'Marek': 9100, 'Kasia': 6300}
print(pensje)
pensje = pd.Series(pensje)
print(pensje)

imie = pensje.idxmax()
print(f'Najwięcej zarabia: {imie}')
print(f'Średnia pensja: {pensje.mean()}')
maska = (pensje > pensje.mean())
print(f'Osoby zarabiające poniżej średniej: \n{pensje[maska]}')
print(f'Poniżej 6000 zarabia {(pensje < 6000).sum()} osób')

{'Anna': 5500, 'Jan': 7200, 'Ewa': 4800, 'Marek': 9100, 'Kasia': 6300}
Anna     5500
Jan      7200
Ewa      4800
Marek    9100
Kasia    6300
dtype: int64
Najwięcej zarabia: Marek
Średnia pensja: 6580.0
Osoby zarabiające poniżej średniej: 
Jan      7200
Marek    9100
dtype: int64
Poniżej 6000 zarabia 2 osób


In [ ]:
podwyzka = pensje * 1.10
print(podwyzka)
premia = pensje + 500
print(premia)
nowePensje = (pensje * 1.10) + 500
print(nowePensje)
print(pensje.sort_values(ascending=False))

Anna      6050.0
Jan       7920.0
Ewa       5280.0
Marek    10010.0
Kasia     6930.0
dtype: float64
Anna     6000
Jan      7700
Ewa      5300
Marek    9600
Kasia    6800
dtype: int64
Anna      6550.0
Jan       8420.0
Ewa       5780.0
Marek    10510.0
Kasia     7430.0
dtype: float64
Marek    9100
Jan      7200
Kasia    6300
Anna     5500
Ewa      4800
dtype: int64


## Ćwiczenie 2
DataFrame - tworzenie i atrybuty

In [2]:
sklep = pd.DataFrame({
'produkt': ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Monitor'],
'cena': [3500, 1800, 2500, 350, 2200],
'stan_magazynowy': [45, 120, 200, 500, 75],
'kategoria': ['Komputery', 'Mobilne', 'Mobilne', 'Akcesoria', 'Komputery']
})
print(sklep)

     produkt  cena  stan_magazynowy  kategoria
0     Laptop  3500               45  Komputery
1     Tablet  1800              120    Mobilne
2   Smartfon  2500              200    Mobilne
3  Słuchawki   350              500  Akcesoria
4    Monitor  2200               75  Komputery


In [9]:
print(sklep.shape)
print(sklep.columns)
print(sklep.dtypes)
print(sklep.head(3))

(5, 6)
Index(['produkt', 'cena', 'stan_magazynowy', 'kategoria', 'wartosc_magazynu',
       'cena_z_wat'],
      dtype='str')
produkt                 str
cena                  int64
stan_magazynowy       int64
kategoria               str
wartosc_magazynu      int64
cena_z_wat          float64
dtype: object
    produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_wat
0    Laptop  3500               45  Komputery            157500      4305.0
1    Tablet  1800              120    Mobilne            216000      2214.0
2  Smartfon  2500              200    Mobilne            500000      3075.0


In [3]:
sklep['wartosc_magazynu'] = sklep['cena'] * sklep['stan_magazynowy']
sklep['cena_z_wat'] = sklep['cena'] * 1.23
print(sklep)

     produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_wat
0     Laptop  3500               45  Komputery            157500      4305.0
1     Tablet  1800              120    Mobilne            216000      2214.0
2   Smartfon  2500              200    Mobilne            500000      3075.0
3  Słuchawki   350              500  Akcesoria            175000       430.5
4    Monitor  2200               75  Komputery            165000      2706.0


In [4]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv'
tips = pd.read_csv(url)

print(tips.head()) 

   total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4


In [7]:
print(tips.shape)
print(tips.tail)
print(tips.sample)
print(tips.dtypes)

(244, 7)
<bound method NDFrame.tail of      total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]>
<bound method NDFrame.sample of      total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinne

## Ćwiczenie 3
 Eksploracja datasetu — samodzielna analiza

In [21]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv'
penguins = pd.read_csv(url)
print(penguins.head())

  species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  


In [22]:
print(penguins.head(10))
print(penguins.shape)
penguins.info()

  species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       36

Zad 3a. #3
Widać 19 brakujących wartości

In [23]:
penguins.describe()

,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


Zad 3b

#1 
Średnia masa osobnika to 4201 gramów, co sugeruje że mamy do czynienia z dużymi ptakami, jednakże
odchylenie standardowe wagi jest duże - 801 gramów - , czyli populacja jest raczej zróżnicowana, są w niej zarówno małe jak i duże ptaki.
Najmniejsze zróżnicowanie występuje w przypaku wysokości dzioba.

In [24]:
penguins.describe(include='all')

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
count,344,344,342.000000,342.000000,342.000000,342.000000,333
unique,3,3,NaN,NaN,NaN,NaN,2
top,Adelie,Biscoe,NaN,NaN,NaN,NaN,MALE
freq,152,168,NaN,NaN,NaN,NaN,168
mean,NaN,NaN,43.921930,17.151170,200.915205,4201.754386,NaN
std,NaN,NaN,5.459584,1.974793,14.061714,801.954536,NaN
min,NaN,NaN,32.100000,13.100000,172.000000,2700.000000,NaN
25%,NaN,NaN,39.225000,15.600000,190.000000,3550.000000,NaN
50%,NaN,NaN,44.450000,17.300000,197.000000,4050.000000,NaN
75%,NaN,NaN,48.500000,18.700000,213.000000,4750.000000,NaN


#2 
Kolumny tekstowe pozwalają odczytać ile mamy unikalnych gatunków lub wysp (unique), która kategoria dominuje (top), oraz jak często występują (freq).

#3
 Mediana masy ciała to 4050 gramów.

In [27]:
print(penguins['species'].value_counts())
print(penguins['island'].value_counts())
print(penguins['sex'].value_counts())
print(penguins['species'].value_counts(normalize = True))

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64
island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64
sex
MALE      168
FEMALE    165
Name: count, dtype: int64
species
Adelie       0.441860
Gentoo       0.360465
Chinstrap    0.197674
Name: proportion, dtype: float64


Zad. 3c

#1 Adelie - 152, Gentoo - 124, Chinstrap - 68 

#2 Biscoe - 168, Dream - 124, Torgersen - 52


#3 Osobników męskich jest 168, żeńskich - 165, czyli rozkład płci jest zbilansowany

#4 Gatunek Adelie stanowi 44,19% badanego zbioru

In [29]:
print(penguins.isna().sum())
braki_plec = penguins['sex'].isna().sum()
procent_brakow = (braki_plec / len(penguins)) * 100
print(procent_brakow)

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64
3.1976744186046515


### Zad. 3e

- Data set zawiera 344 pingwiny z 3 gatunków.
- Najczęstszy gatunek to Adelie. Stanowi on 44,19% wszystkich pingwinów.
- Odchylenie standardowe masy ciała wynosi 801,95g, co przy średniej 4201,75g wskazuje na duże zróżnicowanie badanego zbioru.
- Zbiór jest prawie idealnie zrównoważony pod względem płci, róznica między liczbą samców i samic to zaledwie 3 osobniki.
- Najwięcej osobników (168) pochodzi z wyspy Biscoe, co stanowi prawię połowę całego zbioru.
- Największe braki danych zawiera kolumna określąca płeć, jest to 11 brakujących wartości, co stanowi około 3,2% wszystkich rekordów.

## Ćwiczenie 4
 Pytania biznesowe

In [30]:
tips = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv')
tips['tip_pct'] = (tips['tip'] / tips['total_bill'] * 100).round(1)

In [35]:
print(tips['total_bill'].mean())
print(tips['tip_pct'].mean())
print(tips['day'].value_counts())
print(tips['smoker'].value_counts())
print(tips['tip'].describe())
print(tips['time'].value_counts(normalize= True))
print(tips['size'].nunique())


19.78594262295082
16.08155737704918
day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64
smoker
No     151
Yes     93
Name: count, dtype: int64
count    244.000000
mean       2.998279
std        1.383638
min        1.000000
25%        2.000000
50%        2.900000
75%        3.562500
max       10.000000
Name: tip, dtype: float64
time
Dinner    0.721311
Lunch     0.278689
Name: proportion, dtype: float64
6


### Zad 4a

- Średni rachunek to 19,79.
- Napiwek stanowi średnio 16,1% rachunku.
- Dzień który ma najwięcej rachunków to sobota.
- Rachunków od palaczy jest 93.
- Mediana napiwków to 2,90.
- Z obiadu pochodzi 72,13% rachunków.
- W danych jest 6 unikalnych rozmiarów grup.